### PR TITLE
[SPARK-10960][SQL] SQL with windowing function should be able to refer column in inner select

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-10960

When accessing a column in inner select from a select with window function, `AnalysisException` will be thrown. For example, an query like this:

```
 select area, rank() over (partition by area order by tmp.month) + tmp.tmp1 as c1 from (select month, area, product, 1 as tmp1 from windowData) tmp
```

Currently, the rule `ExtractWindowExpressions` in `Analyzer` only extracts regular expressions from `WindowFunction`, `WindowSpecDefinition` and `AggregateExpression`. We need to also extract other attributes as the one in `Alias` as shown in the above query.

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan  can you take a look at this?

**Comment by cloud-fan** [comment]
Hi @viirya , can you add more description about the reason of this bug? e.g. why does this bug exist and how do you handle it, so that other people can understand your fix easily. Thanks!

**Comment by viirya** [comment]
@cloud-fan Thanks for suggesting. I've updated it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -831,6 +838,13 @@ class Analyzer(
             val withName = Alias(agg, s"_w${extractedExprBuffer.length}")()
             extractedExprBuffer += withName
             withName.toAttribute
+
+          case ne: Alias if hasWindowFunction(ne) && !hasAggregation(ne) =>
+            ne.children.map(_.transform {
+              case e: NamedExpression => extractExpr(e)
+            })
+            ne
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: A one-line change can fix this and pass your test: `case attr: Attribute => extractExpr(attr)`

I think the missing case is `Attribute`, not `Alias`.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Thanks. I've updated it according to your suggestion.

**Comment by viirya** [comment]
ping @cloud-fan Any other comments?

**Comment by cloud-fan** [comment]
cc @yhuai to take another look.

**Comment by yhuai** [comment]
LGTM. @viirya Thank you for fixing this bug.

**Comment by yhuai** [comment]
Merging to master and 1.5 branch.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 2]
```diff
@@ -833,6 +833,33 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+  test("window function: refer column in inner select block") {
+    val data = Seq(
+      WindowData(1, "a", 5),
+      WindowData(2, "a", 6),
+      WindowData(3, "b", 7),
+      WindowData(4, "b", 8),
+      WindowData(5, "c", 9),
+      WindowData(6, "c", 10)
+    )
+    sparkContext.parallelize(data).toDF().registerTempTable("windowData")
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: btw, you can use `withTempTable` in future to let our test infra automatically drop the temp table created at here.

Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Ah, thanks for suggesting it.

**Pull request closed without merging**